packages/std/src/ibc.rs

```diff
@@ -213,37 +211,136 @@ impl IbcAcknowledgement {
+        Self::OpenTry {
+            channel,
+            counterparty_version: counterparty_version.into(),
+        }
+    }
+
+    pub fn channel(&self) -> &IbcChannel {
+        match self {
+            Self::OpenInit { channel } => channel,
+            Self::OpenTry { channel, .. } => channel,
+        }
+    }
+
+    pub fn counterparty_version(&self) -> Option<&str> {
+        match self {
+            Self::OpenTry {
+                counterparty_version,
+                ..
+            } => Some(counterparty_version),
+            _ => None,
+        }
+    }
+}
+
+impl From<IbcChannelOpenMsg> for IbcChannel {
+    fn from(msg: IbcChannelOpenMsg) -> IbcChannel {
+        match msg {
+            IbcChannelOpenMsg::OpenInit { channel } => channel,
+            IbcChannelOpenMsg::OpenTry { channel, .. } => channel,
+        }
     }
 }
 
 /// The message that is passed into `ibc_channel_connect`
 #[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
-pub struct IbcChannelConnectMsg {
-    pub channel: IbcChannel,
+pub enum IbcChannelConnectMsg {
+    OpenAck {
+        channel: IbcChannel,
+        counterparty_version: String,
+    },
+    OpenConfirm {
+        channel: IbcChannel,
+    },
 }
 
 impl IbcChannelConnectMsg {
-    pub fn new(channel: IbcChannel) -> Self {
-        Self { channel }
+    pub fn new_ack(channel: IbcChannel, counterparty_version: impl Into<String>) -> Self {
+        Self::OpenAck {
+            channel,
+            counterparty_version: counterparty_version.into(),
+        }
+    }
+
+    pub fn new_confirm(channel: IbcChannel) -> Self {
+        Self::OpenConfirm { channel }
+    }
+
+    pub fn channel(&self) -> &IbcChannel {
+        match self {
+            Self::OpenAck { channel, .. } => channel,
+            Self::OpenConfirm { channel } => channel,
+        }
+    }
+
+    pub fn counterparty_version(&self) -> Option<&str> {
+        match self {
+            Self::OpenAck {
+                counterparty_version,
+                ..
+            } => Some(counterparty_version),
+            _ => None,
+        }
+    }
+}
+
+impl From<IbcChannelConnectMsg> for IbcChannel {
+    fn from(msg: IbcChannelConnectMsg) -> IbcChannel {
+        match msg {
+            IbcChannelConnectMsg::OpenAck { channel, .. } => channel,
+            IbcChannelConnectMsg::OpenConfirm { channel } => channel,
+        }
     }
 }
 
 /// The message that is passed into `ibc_channel_close`
 #[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
-pub struct IbcChannelCloseMsg {
-    pub channel: IbcChannel,
+pub enum IbcChannelCloseMsg {
+    CloseInit { channel: IbcChannel },
+    CloseConfirm { channel: IbcChannel }, // pub channel: IbcChannel,
 }
 
 impl IbcChannelCloseMsg {
-    pub fn new(channel: IbcChannel) -> Self {
-        Self { channel }
+    pub fn new_init(channel: IbcChannel) -> Self {
```

Comment:
Thoughts about these constructors?

Reply:
look good to me, but I'll leave naming decisions to you and Simon

Reply:
LGTM